src/lib/client/proving.ts

```diff
@@ -0,0 +1,325 @@
+  const requiredSigPubKeysEdwards = requiredSigPubKeys.map((pubKey) =>
+    publicKeyFromString(pubKey).toEdwards()
+  );
+  const proofSigUsers = collectedSigUsers.slice(0, requirement.numSigsRequired);
+
+  if (onUpdateProvingState) {
+    onUpdateProvingState({
+      currentRequirementNumSigsTotal: requirement.numSigsRequired,
+      currentRequirementNumSigsProven: 0,
+    });
+  }
+
+  const proofs: MembershipProof[] = [];
+  for (let i = 0; i < proofSigUsers.length; i++) {
+    const user = proofSigUsers[i];
+    const sig = derDecodeSignature(user.sig!);
+    const msgHash = hexToBigInt(getECDSAMessageHash(user.msg!));
+    const index = requiredSigPubKeys.indexOf(user.sigPk!);
+    const pubKey = publicKeyFromString(user.sigPk!);
+    const publicInputs = getPublicInputsFromSignature(sig, msgHash, pubKey);
+    const merkleProof = await computeMerkleProof(
+      requiredSigPubKeysEdwards,
+      index,
+      hashFn
+    );
+
+    const proof = await proveMembership({
+      sig,
+      msgHash,
+      publicInputs,
+      merkleProof,
+      sigNullifierRandomness,
+      pubKeyNullifierRandomness,
+      pathToCircuits: getClientPathToCircuits(),
+    });
+    proofs.push(proof);
+
+    if (onUpdateProvingState) {
+      onUpdateProvingState({
+        currentRequirementNumSigsTotal: requirement.numSigsRequired,
+        currentRequirementNumSigsProven: i + 1,
+      });
+    }
+  }
+
+  return proofs;
+};
+
+const generateProofForLocationRequirement = async (
+  locations: Record<string, LocationSignature>,
+  requirement: LocationRequirement,
+  hashFn: any,
+  onUpdateProvingState?: (newProvingState: {
+    currentRequirementNumSigsTotal: number;
+    currentRequirementNumSigsProven: number;
+  }) => void
+): Promise<MembershipProof[]> => {
+  const requiredSigPubKeys = requirement.locations.map(
+    (location) => location.signaturePublicKey
+  );
+  const collectedSigLocations = Object.values(locations).filter(
+    (location) =>
+      location.pk &&
+      location.sig &&
+      location.msg &&
+      requiredSigPubKeys.includes(location.pk)
+  );
+  if (collectedSigLocations.length < requirement.numSigsRequired) {
+    throw new Error("Not enough signatures for location requirement");
+  }
+
+  const sigNullifierRandomness = hexToBigInt(
+    requirement.sigNullifierRandomness
+  );
+  const pubKeyNullifierRandomness = hexToBigInt(
+    getClientRandomNullifierRandomness()
+  ); // Ensures user cannot reuse a public key in a proof for this requirement
+  const requiredSigPubKeysEdwards = requiredSigPubKeys.map((pubKey) =>
+    publicKeyFromString(pubKey).toEdwards()
+  );
+  const proofSigLocations = collectedSigLocations.slice(
+    0,
+    requirement.numSigsRequired
+  );
+
+  if (onUpdateProvingState) {
+    onUpdateProvingState({
+      currentRequirementNumSigsTotal: requirement.numSigsRequired,
+      currentRequirementNumSigsProven: 0,
+    });
+  }
+
+  const proofs: MembershipProof[] = [];
```

Comment:
We should make an issue to try this with web workers, depending on how many requirements certain quests have.

Reply:
https://github.com/nfc-denver/nfc-denver/issues/85